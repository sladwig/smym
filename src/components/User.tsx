import React from 'react';
import { IUser } from '../store/User';
import { observer } from 'mobx-react-lite';

interface IProps {
    user: IUser;
}

function User({ user }: IProps) {
    return <>{user.name}</>;
}

export default observer(User);
