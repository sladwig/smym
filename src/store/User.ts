import { types, Instance } from 'mobx-state-tree';
import { Transaction } from './Transaction';

export const User = types
    .model('User', {
        id: types.identifier,
        name: types.maybe(types.string),
        real_name: types.maybe(types.string),
        dmChannel: types.maybe(types.string),
        transactions: types.array(Transaction),
    })
    .actions(self => ({
        setDmChannel(channelId: string) {
            self.dmChannel = channelId;
        },
        addTransaction(value: number | 'reset', description: string) {
            if (value === 'reset') return this.resetTransactions();
            self.transactions.push({ value, description });
        },
        resetTransactions() {
            self.transactions.clear();
        },
        update(user: any) {
            self.name = user.name;
            self.real_name = user.real_name;
        },
    }))
    .views(self => ({
        get balance() {
            return self.transactions.reduce((sum, current) => sum + current.value, 0);
        },
    }));

export type IUser = Instance<typeof User>;
