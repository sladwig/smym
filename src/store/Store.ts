import { types, SnapshotOrInstance } from 'mobx-state-tree';
import { User, IUser } from './User';

export const Store = types
    .model('Store', {
        users: types.map(User),
    })
    .actions(self => ({
        addUser(user: SnapshotOrInstance<typeof User>) {
            self.users.put(user);
        },
    }))
    .views(self => ({
        get usersList(): IUser[] {
            return Array.from(self.users.values());
        },
    }));

export const store = Store.create({});
