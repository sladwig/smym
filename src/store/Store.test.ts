import { Store } from './Store';
import { IUser } from './User';
import { getSnapshot } from 'mobx-state-tree';

it('can be created', () => {
    const store = Store.create({});
    expect(store).toMatchSnapshot();
});

it('can add user', () => {
    const store = Store.create();
    expect(store.users.size).toEqual(0);

    store.addUser({ id: 'def', name: 'mark' });
    expect(store.users.size).toEqual(1);

    const user = store.users.get('def');
    expect(user.name).toEqual('mark');
});

it('has a userList', () => {
    const store = Store.create();
    store.addUser({ id: 'def', name: 'mark' });
    expect(store.usersList).toMatchSnapshot();
    expect(store.usersList.length).toEqual(1);
});
