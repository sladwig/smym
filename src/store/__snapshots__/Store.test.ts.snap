// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can be created 1`] = `
Object {
  "users": Object {},
}
`;

exports[`has a userList 1`] = `
Array [
  Object {
    "dmChannel": undefined,
    "id": "def",
    "name": "mark",
    "real_name": undefined,
    "transactions": Array [],
  },
]
`;
